### PR TITLE
[RFC]rats-tls: add tdx event log support

### PR DESCRIPTION
**Pull request by YangLiang3**
Fixes: #31
Signed-off-by: Liang Yang <liang3.yang@intel.com>

### CODE REVIEW

File: src/attesters/tdx-ecdsa/collect_evidence.c [hunk 1]
```diff
@@ -35,6 +42,47 @@ static int tdx_get_report(const tdx_report_data_t *report_data, tdx_report_t *td
...
+    int fd = 0;
+    char tdel_info[TDEL_INFO_SZ];
+
+    RTLS_DEBUG("ctx %p, evidence %p\n", ctx, evidence);
+    if (fd = open(TDEL_INFO, O_RDONLY) < 0) {
+        RTLS_ERR("failed to open TDEL info device\n");
+        return -ENCLAVE_ATTESTER_ERR_INVALID;
+    }
+
+    int tdel_info_sz = read(fd, tdel_info, TDEL_INFO_SZ);
```
Comment (jiazhang0) [inline src/attesters/tdx-ecdsa/collect_evidence.c:56]: You forget to close fd.

File: src/attesters/tdx-ecdsa/collect_evidence.c [hunk 2]
```diff
@@ -35,6 +42,47 @@ static int tdx_get_report(const tdx_report_data_t *report_data, tdx_report_t *td
...
+    if (tdel_info_sz != TDEL_INFO_SZ) {
```
Comment (jiazhang0) [inline src/attesters/tdx-ecdsa/collect_evidence.c:57]: How do you know the size of event log is exactly 0x10000 bytes?

File: src/attesters/tdx-ecdsa/collect_evidence.c [hunk 3]
```diff
@@ -35,6 +42,47 @@ static int tdx_get_report(const tdx_report_data_t *report_data, tdx_report_t *td
...
+
+    RTLS_DEBUG("ctx %p, evidence %p\n", ctx, evidence);
+    if (fd = open(TDEL_DATA, O_RDONLY) < 0) {
+        RTLS_ERR("failed to open TDEL info device\n");
+        return -ENCLAVE_ATTESTER_ERR_INVALID;
+    }
+
+    int tdel_data_sz = read(fd, tdel_data, TDEL_DATA_SZ);
+
+    memcpy(&(evidence->tdx.quote[8192 + TDEL_INFO_SZ]), tdel_data, tdel_data_sz);
```
Comment (jiazhang0) [inline src/attesters/tdx-ecdsa/collect_evidence.c:81]: 8192 should be replaced by TDX_ECDSA_QUOTE_SZ?

File: src/attesters/tdx-ecdsa/collect_evidence.c [hunk 4]
```diff
@@ -35,6 +42,47 @@ static int tdx_get_report(const tdx_report_data_t *report_data, tdx_report_t *td
...
+        return -ENCLAVE_ATTESTER_ERR_INVALID;
+    }
+
+    int tdel_info_sz = read(fd, tdel_info, TDEL_INFO_SZ);
+    if (tdel_info_sz != TDEL_INFO_SZ) {
+        RTLS_ERR("failed to read tdinfo\n");
+        return -ENCLAVE_ATTESTER_ERR_INVALID;
+    }
+
+    memcpy(&(evidence->tdx.quote[8192]), tdel_info, TDEL_INFO_SZ);
```
Comment (jiazhang0) [inline src/attesters/tdx-ecdsa/collect_evidence.c:62]: 8192 should be TDX_ECDSA_QUOTE_SZ?

File: src/attesters/tdx-ecdsa/collect_evidence.c [hunk 5]
```diff
@@ -35,6 +42,47 @@ static int tdx_get_report(const tdx_report_data_t *report_data, tdx_report_t *td
...
+    char tdel_data[TDEL_DATA_SZ];
+
+    RTLS_DEBUG("ctx %p, evidence %p\n", ctx, evidence);
+    if (fd = open(TDEL_DATA, O_RDONLY) < 0) {
+        RTLS_ERR("failed to open TDEL info device\n");
+        return -ENCLAVE_ATTESTER_ERR_INVALID;
+    }
+
+    int tdel_data_sz = read(fd, tdel_data, TDEL_DATA_SZ);
+
```
Comment (jiazhang0) [inline src/attesters/tdx-ecdsa/collect_evidence.c:80]: fd should be closed.

File: src/attesters/tdx-ecdsa/collect_evidence.c (hunk 2 above)
Comment (jiazhang0) [reply src/attesters/tdx-ecdsa/collect_evidence.c:57]: So you miss to parse tdel_info which can tell the exact size of event log to read. You can consult Liang Zhou for the details.

File: src/attesters/tdx-ecdsa/collect_evidence.c [hunk 6]
```diff
@@ -86,15 +134,20 @@ enclave_attester_err_t tdx_ecdsa_collect_evidence(enclave_attester_ctx_t *ctx,
...
 
-	evidence->tdx.quote_len = sizeof(evidence->tdx.quote);
-	if (tdx_gen_quote(hash, evidence->tdx.quote, &evidence->tdx.quote_len)) {
-		RTLS_ERR("failed to generate quote\n");
-		return -ENCLAVE_ATTESTER_ERR_INVALID;
-	}
+    if (tdx_gen_quote(hash, evidence->tdx.quote, &evidence->tdx.quote_len)) {
+        RTLS_ERR("failed to generate quote\n");
+        return -ENCLAVE_ATTESTER_ERR_INVALID;
+    }
```
Comment (jiazhang0) [inline src/attesters/tdx-ecdsa/collect_evidence.c:142]: Lots of format issue causes such changes.

File: src/include/rats-tls/cert.h [hunk 7]
```diff
@@ -7,6 +7,11 @@
 #ifndef _ENCLAVE_CERT_H
 #define _ENCLAVE_CERT_H
 
+#define SGX_ECDSA_QUOTE_SZ 8192
+#define TDX_ECDSA_QUOTE_SZ 8192
+#define TDEL_INFO_SZ      56
+#define TDEL_DATA_SZ      65536
```
Comment (jiazhang0) [inline src/include/rats-tls/cert.h:13]: I don't think you need to define these macros again.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: src/include/rats-tls/cert.h [hunk 8]
```diff
@@ -35,7 +40,7 @@ typedef struct {
 } la_attestation_evidence_t;
 
 typedef struct {
-	uint8_t quote[8192];
+    uint8_t quote[TDX_ECDSA_QUOTE_SZ + TDEL_INFO_SZ + TDEL_DATA_SZ];
 	uint32_t quote_len;
```
Comment (haosanzi) [inline src/include/rats-tls/cert.h:44]: Please fix the format issue.

**Pull request closed without merging**